### PR TITLE
feat: add agent triggers

### PR DESCRIPTION
**Pull request by qinxuye**
## Summary
- add agent-bound webhook and scheduled trigger persistence, API routes, and run history
- keep scheduled trigger scans in Celery as prepare-only work and dispatch agent execution from backend processes
- add trigger run status synchronization when hidden trigger tasks finish

## Validation
- pytest tests/web/api/test_agent_triggers.py tests/web/test_background_jobs.py::test_trigger_event_job_runs_with_eager_celery tests/web/test_background_jobs.py::test_trigger_event_idempotency_is_scoped_by_user tests/web/api/v1/test_tasks.py::test_create_task_happy_path tests/web/api/v1/test_tasks.py::test_get_task_returns_404_for_non_sdk_source tests/web/api/v1/test_tasks.py::test_append_message_returns_404_for_non_sdk_source tests/web/api/v1/test_tasks.py::test_get_steps_returns_404_for_non_sdk_source tests/core/test_config.py::TestEnvironmentVariableConstants::test_celery_background_job_constants tests/core/test_config.py::TestCeleryBackgroundJobConfig -q
- ruff check src/xagent/web/models/trigger.py src/xagent/web/services/triggers.py src/xagent/web/api/triggers.py src/xagent/web/jobs/trigger_tasks.py src/xagent/web/services/task_orchestrator.py src/xagent/web/app.py src/xagent/config.py tests/web/api/test_agent_triggers.py tests/core/test_config.py src/xagent/migrations/versions/20260616_add_agent_triggers.py
- pre-commit hooks via git commit

### CODE REVIEW

File: src/xagent/web/api/triggers.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+
+async def _read_payload(request: Request) -> dict[str, Any]:
+    body = await request.body()
+    if not body:
+        return {}
+    try:
+        decoded = json.loads(body.decode("utf-8"))
+    except json.JSONDecodeError:
+        return {"body": body.decode("utf-8", errors="replace")}
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply src/xagent/web/api/triggers.py:334]: Fixed in the latest push by catching ValueError for both JSON parsing and UTF-8 decode failures. I also added a regression test for invalid UTF-8 webhook bodies.

File: src/xagent/web/services/triggers.py [hunk 2]
```diff
@@ -15,6 +35,763 @@
...
+        raise TriggerServiceError("interval_seconds must be an integer") from exc
+    if interval_seconds <= 0:
+        raise TriggerServiceError("interval_seconds must be positive")
+
+    candidate = base + timedelta(seconds=interval_seconds)
+    now = from_time or _now()
+    now = _coerce_utc(now) or _now()
+    while candidate <= now:
+        candidate += timedelta(seconds=interval_seconds)
+    return candidate
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply src/xagent/web/services/triggers.py:196]: Fixed in the latest push by replacing the catch-up loop with constant-time arithmetic, so a stale scheduled trigger advances directly to the next future interval.

File: src/xagent/web/services/triggers.py [hunk 3]
```diff
@@ -15,6 +35,763 @@
...
+        db.query(TriggerRun)
+        .join(Task, TriggerRun.task_id == Task.id)
+        .filter(
+            TriggerRun.status == TriggerRunStatus.PENDING.value,
+            Task.status == TaskStatus.PENDING,
+        )
+        .order_by(TriggerRun.created_at.asc(), TriggerRun.id.asc())
+        .limit(max(1, min(limit, 100)))
+        .all()
+    )
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply src/xagent/web/services/triggers.py]: Addressed in the latest push, but not by reusing the caller SQLAlchemy Session inside to_thread. The dispatcher now does the blocking pending-run-id lookup with a thread-local session, then reloads those runs in the caller session before starting them.

File: src/xagent/web/api/triggers.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+        enabled=bool(trigger.enabled),
+        config=dict(trigger.config or {}),
+        prompt_template=trigger.prompt_template,
+        webhook_token=trigger.webhook_token,
+        webhook_secret=webhook_secret,
+        next_run_at=_dt(getattr(trigger, "next_run_at", None)),
+        last_run_at=_dt(getattr(trigger, "last_run_at", None)),
+        last_error=trigger.last_error,
+        created_at=_dt(getattr(trigger, "created_at", None)),
+        updated_at=_dt(getattr(trigger, "updated_at", None)),
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply src/xagent/web/api/triggers.py]: Addressed in the latest push. The serializer now reads trigger.next_run_at / last_run_at / created_at / updated_at directly and casts the old-style SQLAlchemy Column typing for mypy.

File: src/xagent/web/services/task_orchestrator.py [hunk 5]
```diff
@@ -709,6 +716,40 @@ def finish_turn(bg_db: Any, task_id: int) -> None:
     # PAUSED / WAITING_FOR_USER / other: leave alone.
 
 
+def _sync_trigger_run_status(bg_db: Any, task: Task, status: TaskStatus) -> bool:
+    """Best-effort mirror from task terminal state to trigger run history."""
+    from ..models.trigger import TriggerRun, TriggerRunStatus
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply src/xagent/web/services/task_orchestrator.py:730]: Fixed in the latest push by adding finish_turn_isolated and calling it via asyncio.to_thread from the runner. The helper opens its own short-lived SessionLocal in the worker thread, so finish_turn and _sync_trigger_run_status no longer run their synchronous DB work on the main event loop.

File: src/xagent/web/services/triggers.py [hunk 6]
```diff
@@ -15,6 +35,791 @@
...
+                error_message="Prepared trigger task or trigger is missing",
+            )
+        return False
+
+    if task.status == TaskStatus.RUNNING:
+        setattr(run, "status", TriggerRunStatus.RUNNING.value)
+        setattr(run, "started_at", getattr(run, "started_at", None) or _now())
+        db.add(run)
+        db.commit()
+        return False
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply src/xagent/web/services/triggers.py]: Addressed in the latest push. The read side now uses run.started_at directly; writes stay with setattr because these SQLAlchemy models use old-style Column typing and direct assignment fails mypy in this repo.

File: src/xagent/web/services/triggers.py [hunk 7]
```diff
@@ -15,6 +35,791 @@
...
+    finally:
+        db.close()
+
+
+async def dispatch_pending_trigger_runs(
+    db: Session,
+    *,
+    limit: int = 20,
+    wait_for_completion: bool = False,
+) -> int:
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply src/xagent/web/services/triggers.py:885]: Fixed in the latest push. The dispatcher now only fetches pending run ids off-loop, and each prepared-run DB check/status update runs through worker-thread helpers with their own SessionLocal. The actual TaskTurnOrchestrator.begin_turn call stays in the async flow so task scheduling remains on the backend event loop.

File: src/xagent/web/api/triggers.py [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+            or request.headers.get("x-event-id")
+            or request.headers.get("x-request-id")
+        )
+        run, created = await fire_trigger(
+            db,
+            trigger=trigger,
+            event_payload=payload,
+            source_event_id=source_event_id,
+        )
+        return TriggerFireResponse(
```
Comment (rogercloud) [inline src/xagent/web/api/triggers.py]: **MAJOR (M1):** Public webhook callers should not receive `task_id`, `background_job_id`, `idempotency_key`, or the full `payload_snapshot`. Consider introducing a separate minimal response model (e.g. `trigger_run_id`, `status`, `duplicate`) for this endpoint.

File: src/xagent/web/api/triggers.py [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+
+
+def _handle_service_error(exc: Exception) -> HTTPException:
+    if isinstance(exc, TriggerNotFoundError):
+        return HTTPException(status_code=404, detail=str(exc))
+    if isinstance(exc, TriggerSecretError):
+        return HTTPException(status_code=401, detail=str(exc))
+    if isinstance(exc, TriggerServiceError):
+        return HTTPException(status_code=400, detail=str(exc))
+    return HTTPException(status_code=500, detail=str(exc))
```
Comment (rogercloud) [inline src/xagent/web/api/triggers.py]: **MAJOR (M2):** Returning `detail=str(exc)` for unhandled exceptions can leak internal DB/SQL details. Log the exception and return a generic `"Internal server error"` message for the 500 case.

File: src/xagent/web/services/triggers.py [hunk 10]
```diff
@@ -15,6 +36,880 @@
...
+        "preview": encoded[:_PAYLOAD_PREVIEW_LIMIT],
+    }
+
+
+def _payload_hash(payload: dict[str, Any]) -> str:
+    return hashlib.sha256(_json_dumps(payload).encode("utf-8")).hexdigest()
+
+
+def _hash_secret(secret: str) -> str:
+    return hashlib.sha256(secret.encode("utf-8")).hexdigest()
```
Comment (rogercloud) [inline src/xagent/web/services/triggers.py]: **MAJOR (M3):** Storing webhook secrets as a plain SHA-256 hash is too fast for low-entropy user secrets. Since this feature is new and has no existing hashes to migrate, please use `bcrypt` (already a project dependency) for `_hash_secret` / `verify_webhook_secret`.

File: src/xagent/web/api/triggers.py [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+)
+
+logger = logging.getLogger(__name__)
+
+router = APIRouter(tags=["triggers"])
+
+
+class TriggerCreateRequest(BaseModel):
+    type: Literal["webhook", "scheduled"]
+    name: str | None = None
```
Comment (rogercloud) [inline src/xagent/web/api/triggers.py]: **MAJOR (M4):** `name` has no `max_length` or non-empty validation. The DB column is `String(200)`, so oversized names produce a 500 `DataError`; empty/whitespace-only names can also be persisted. Add `Field(max_length=200)` here and reject empty names in the service layer.

File: src/xagent/web/services/triggers.py [hunk 12]
```diff
@@ -15,6 +36,880 @@
...
+
+async def dispatch_pending_trigger_runs(
+    db: Session,
+    *,
+    limit: int = 20,
+    wait_for_completion: bool = False,
+) -> int:
+    """Start prepared trigger tasks from the backend process."""
+    pending_run_ids = await asyncio.to_thread(
+        _get_pending_trigger_run_ids,
```
Comment (rogercloud) [inline src/xagent/web/services/triggers.py:888]: **MAJOR (M5):** This is a check-then-act race. `_get_pending_trigger_run_ids` reads pending runs in one session and then each run is started in another session/transaction. Under multiple backend workers the same `TriggerRun` can be picked up concurrently, leading to redundant start attempts and conflicting state updates. Use an atomic claim such as `UPDATE trigger_runs SET status='running', started_at=... WHERE status='pending' AND id=:id RETURNING id` so only one worker can claim each run.

**Review by rogercloud — CHANGES_REQUESTED** [review]
## Summary

Thanks for the feature. I reviewed the latest head (`3d52db6`) against `main` and ran the PR's specified test set plus the targeted `ruff check`.

- **pytest:** 23 passed, 0 failed ✅
- **ruff check:** clean ✅

However, there are **5 major issues** that should be addressed before this is merged, mostly around security, API contract, and multi-worker correctness.

## Major issues

**M1. Public webhook response leaks internal IDs** — `src/xagent/web/api/triggers.py:363-365`
The public webhook endpoint returns the full `TriggerRunResponse`, including `task_id`, `background_job_id`, `idempotency_key`, and `payload_snapshot`. External callers only need a minimal acknowledgement.

**M2. 500 errors expose raw exception strings** — `src/xagent/web/api/triggers.py:162-169`
`_handle_service_error` returns `detail=str(exc)` for unhandled exceptions, which can leak DB/SQL details.

**M3. Webhook secrets are stored as plain SHA-256** — `src/xagent/web/services/triggers.py:97-104`
`_hash_secret` uses a fast SHA-256 digest. User-supplied webhook secrets may be low-entropy, so this is vulnerable to offline brute force if the DB is leaked. Since this is a new feature with no existing secrets, prefer `bcrypt`.

**M4. Trigger `name` has no length/empty validation** — `src/xagent/web/api/triggers.py:35-56` and `src/xagent/web/services/triggers.py:197-292`
`name` lacks `max_length` and empty/whitespace checks, allowing oversized names to hit a 500 `DataError` and empty names to be persisted.

**M5. Dispatcher has a check-then-act race across workers** — `src/xagent/web/services/triggers.py:802-846`
`_get_pending_trigger_run_ids` reads pending runs without locking, and `_start_prepared_trigger_run_id` acts on them in separate sessions. Multiple backend workers can race on the same `TriggerRun`, causing redundant start attempts and noisy/incorrect state transitions.

## Minor issues and test gaps

- `trigger.last_error` is not updated when a started run fails (`task_orchestrator.py:719-750`).
- Test runs still update `trigger.last_run_at` (`triggers.py`).
- `list_trigger_runs` is hard-capped at 100 with no pagination.
- Non-dict JSON payloads are dropped from `_serialize_run`.
- Migration adds redundant indexes on primary-key columns.
- Missing tests for: cross-user authorization, negative input validation, delete endpoint, failure-after-start path, and the Celery `enqueue_trigger_event_job` path.

## Recommendation

Please address the 5 major issues above. Happy to re-review once updated.

File: src/xagent/web/api/triggers.py (hunk 11 above)
Comment (qinxuye) [reply src/xagent/web/api/triggers.py]: Fixed in the latest push. Create/update schemas now cap trigger names at 200 characters, and the service rejects empty or whitespace-only names before hitting the database.

File: src/xagent/web/services/triggers.py (hunk 10 above)
Comment (qinxuye) [reply src/xagent/web/services/triggers.py]: Fixed in the latest push. Webhook secrets now use bcrypt with the same cost factor used by the existing API key code, and verification uses bcrypt.checkpw with malformed hashes treated as auth failure.

File: src/xagent/web/api/triggers.py (hunk 8 above)
Comment (qinxuye) [reply src/xagent/web/api/triggers.py]: Fixed in the latest push. The public webhook endpoint now returns a separate minimal response with trigger_run_id, status, and duplicate only; it no longer exposes task_id, background_job_id, idempotency_key, or payload_snapshot.

File: src/xagent/web/services/triggers.py (hunk 12 above)
Comment (qinxuye) [reply src/xagent/web/services/triggers.py:888]: Fixed in the latest push. Each pending run now has to win an atomic conditional UPDATE from pending to running before any task start work proceeds; concurrent workers that lose the rowcount check skip that run. I added a concurrent-start regression test covering this path.

File: src/xagent/web/api/triggers.py (hunk 9 above)
Comment (qinxuye) [reply src/xagent/web/api/triggers.py]: Fixed in the latest push. Unexpected trigger API errors are logged server-side and now return a generic Internal server error detail instead of str(exc).

**Review by rogercloud — APPROVED** [review]
## Re-review: all major issues resolved

Thanks for the quick turnaround. I re-checked the latest head (`a4a2d9d`) and all 5 previously requested changes are now addressed:

- **M1.** Public webhook now returns `PublicTriggerFireResponse` with only `trigger_run_id`, `status`, and `duplicate`.
- **M2.** `_handle_service_error` logs unhandled exceptions and returns a generic `"Internal server error"` for 500s.
- **M3.** Webhook secrets are now hashed with `bcrypt` (cost 12).
- **M4.** `TriggerCreateRequest` / `TriggerUpdateRequest` enforce `max_length=200`, and `_normalize_trigger_name` rejects empty/whitespace names.
- **M5.** Dispatcher now atomically claims pending runs via `_claim_pending_trigger_run` (`UPDATE ... WHERE status='pending'`) before starting them.

## Validation

- **pytest:** 25 passed, 0 failed ✅
- **ruff check:** clean ✅

The remaining minor items (pagination on run history, redundant migration indexes, a few test gaps) are non-blocking and can be addressed in follow-up work.

Approving.
